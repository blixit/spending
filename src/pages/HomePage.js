import React from 'react';
import styled from 'styled-components';
import trigger from 'vanillajs-browser-helpers/trigger';
import { FaTrash } from 'react-icons/fa';

import ThemedButton from 'components/atoms/buttons/ThemedButton';
import FilterForm from 'components/forms/filter-form/FilterForm';

import {
  Get,
  ReadyQueries as Queries,
  mutate,
  ReadyMutations as Mutations
} from 'core/http/query';

import PageComponent from 'pages/Page';

const Page = styled(PageComponent)`
  flex-direction: column;
`;

const Label = styled.span`
  font-size: 1.3em;
`;

const Price = styled.span`
  font-size: 1.215em;
`;

const Item = styled.div`
  display: flex;
  justify-content: space-between;
  padding-top: 10px;
  padding-bottom: 10px;
  font-size: 1.2em;
  border-bottom: 1px solid silver;
`;

const StyledFilterForm = styled(FilterForm)`
  margin-top: ${({ showFilter }) => showFilter ? '20px' : 0 };
  margin-bottom: 20px;
`;

const Trash = styled(FaTrash)`
  margin-left: 10px;
  cursor: ${({ onClick }) => onClick ? 'pointer' : 'default' };
`;

class HomePage extends React.Component {
  static displayName = 'HomePage';

  state = {
    showFilter: true,
    items: []
  };

  toggleFilter = () => {
    const { showFilter } = this.state;
    this.setState({ showFilter: !showFilter });
  };

  formatDate = (date) => {
    const toFormat = new Date(date);
    return toFormat.toLocaleString();
  };

  onTrashClicked = ({ id }) => {
    const { items } = this.state;
    const { remove } = Mutations.spending;
    mutate({ ...remove, data: {
      id
    }}).then(data => {
      const filtered = items.filter(item => item.id !== id);
      this.setState({ items: filtered });
      trigger(document, 'refresh:balance', data);
    }).catch(error => {
      console.error(error);
    });
  };

  renderItem = (item) => {
    const { id, label, date, price } = item;
    return (
      <Item key={`${id}`}>
        <div>
          <Label>{label}</Label><br />
          <span>{this.formatDate(date)}</span>
        </div>
        <div>
          <Price>{price} €</Price>
          <Trash onClick={() => this.onTrashClicked(item)} />
        </div>
      </Item>
    );
  };

  renderItems = (data) => {
    return (data || []).map(
      (item) => this.renderItem(item)
    );
  };

  updateItems = ({ status, data: { data: items } }) => {
    this.setState({ items });
  };

  render () {
    const { showFilter, items } = this.state;
    const { ...rest } = this.props;

    return (
      <Page {...rest}>
        <React.Fragment>
          <ThemedButton color='primary' onClick={this.toggleFilter}>Recherche avancée</ThemedButton>
          {showFilter && <StyledFilterForm
            showFilter={showFilter}
            onFormResult={this.updateItems}
          />}
        </React.Fragment>

        <Get
          {...Queries.test}
          children={this.updateItems}
        />
        {this.renderItems(items)}
      </Page>
    );
  }
}

export default HomePage;
