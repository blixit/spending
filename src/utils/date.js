import moment from 'moment';

export const locale = 'fr-FR';

const getDate = date =>
  moment(date)
  .format(process.env.REACT_APP_FORMATS_DATE);

const getTime = date =>
  moment(date)
  .format(process.env.REACT_APP_FORMATS_TIME);

/**
 * Formats a date for the backend.
 * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleDateString
 * @param {Date} date the date to convert
 */
export const toBackendDate = date => {
  return date
    ? (getDate(date) + ' ' + getTime(date))
    : date;
};

export const strToDatetime = toFormat => {
  const date = new Date(toFormat);
  return toBackendDate(date);
};
