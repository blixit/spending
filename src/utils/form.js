/**
* Removes the listed fields from the source
* @param {*} form The source object
* @param {*} fields The fields list
*/
export const clean = (form, fields) => {
  const result = {};
  // eslint-disable-next-line no-unused-vars
  for (const field in form) {
    if (!form.hasOwnProperty(field) || fields.includes(field)) {
      continue;
    }
    result[field] = form[field];
  }
  return result;
};
