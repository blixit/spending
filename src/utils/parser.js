export const stringify = data => {
  try {
    return JSON.stringify(data);
  } catch (error) {
    return null;
  };
}

export const parse = data => JSON.parse(data);