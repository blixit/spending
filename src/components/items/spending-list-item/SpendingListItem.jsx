import React, { useCallback } from 'react';
import styled from 'styled-components';
import { FaTrash } from 'react-icons/fa';

import { strToDatetime } from 'utils/date';

const Label = styled.span`
`;
const Date = styled.span`
  font-size: 0.7rem;
  font-style: italic;
`;

const Price = styled.span`
  font-size: 1.15rem;
`;

const PriceContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
`;

const Item = styled.div`
  display: flex;
  justify-content: space-between;
  margin-top: 10px;
  margin-bottom: 10px;
  padding: 10px;
  background-color: ${({ theme }) => theme.white()}
  font-size: 1.2em;
`;

const Trash = styled(FaTrash)`
  margin-left: 10px;
  cursor: ${({ onClick }) => onClick ? 'pointer' : 'default' };
`;


const SpendingListItem = ({ item, onDelete }) => {
  const { label, date, price } = item;

  const onDeleteItem = useCallback(() => {
    onDelete && onDelete(item);
  }, [item, onDelete]);

  return (
    <Item>
      <div>
        <Label>{label}</Label><br />
        <Date>{strToDatetime(date)}</Date>
      </div>
      <PriceContainer>
        <Price>{price} â‚¬</Price>
        <Trash onClick={onDeleteItem} />
      </PriceContainer>
    </Item>
  );
};

export default SpendingListItem;
