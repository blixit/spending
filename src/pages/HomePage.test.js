import React from 'react';
import { act } from 'react-test-renderer';

import ThemeProvider from 'core/theming/theme';
import { withTheme } from 'core/theming/test.utils';

import SpendingListItem from 'components/items/spending-list-item/SpendingListItem';

import render from 'testing/render';

import { HomePage } from './HomePage';

describe('[pages] HomePage', () => {
  const getInstance = (tree) => tree.root
    .findByType(ThemeProvider)
    .children[0].children[0];
  const items = {
    data: {
      data: [{
        id: 27,
        label: 'My depense',
        price: 45,
        date: '2019-09-25T23:38:53+02:00'
      }]
    }
  };

  it('should render properly', () => {
    const tree = render(withTheme(<HomePage items={items} />));
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should render items when provided', () => {
    const tree = render(withTheme(<HomePage items={items} />));
    const listItems = tree.root.findAllByType(SpendingListItem);
    expect(listItems).toHaveLength(items.data.data.length);
  });

  it('change state when toggleFilter is called', () => {
    const tree = render(withTheme(<HomePage />));
    
    let homepage = getInstance(tree);
    act(() => {
      homepage.instance.toggleFilter();
    });
    tree.update(withTheme(<HomePage />));
    expect(homepage.instance.state.showFilter).toBe(true);

    act(() => {
      homepage.instance.toggleFilter();
    });
    expect(homepage.instance.state.showFilter).toBe(false);
  });
});
