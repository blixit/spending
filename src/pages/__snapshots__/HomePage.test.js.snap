// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[pages] HomePage should render properly 1`] = `
<div
  className="sc-ckVGcZ jbCulJ sc-hMqMXs bYupWY"
>
  <div
    className="sc-jKJlTe hvKJPV"
  >
    <button
      className="sc-bZQynM hGfsCj sc-gzVnrw fIeHAv sc-htoDjs eFJrQB sc-eNQAEJ ipBAUA"
      onClick={[Function]}
    >
      <svg
        color="white"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": "white",
          }
        }
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M487.976 0H24.028C2.71 0-8.047 25.866 7.058 40.971L192 225.941V432c0 7.831 3.821 15.17 10.237 19.662l80 55.98C298.02 518.69 320 507.493 320 487.98V225.941l184.947-184.97C520.021 25.896 509.338 0 487.976 0z"
        />
      </svg>
    </button>
    <span
      className="sc-kkGfuU kWebY"
    >
      Your depenses:
    </span>
    <div>
      <div
        className="sc-ifAKCX ddPLTk"
      >
        <div>
          <span
            className="sc-bdVaJa fePDys"
          >
            My depense
          </span>
          <br />
          <span
            className="sc-bwzfXH eYuQBz"
          >
            25/09/2019 21:38:53
          </span>
        </div>
        <div
          className="sc-bxivhb jtHuoG"
        >
          <span
            className="sc-htpNat eSrccS"
          >
            45
             â‚¬
          </span>
          <svg
            className="sc-EHOje bvPvHy"
            fill="currentColor"
            height="1em"
            onClick={[Function]}
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
